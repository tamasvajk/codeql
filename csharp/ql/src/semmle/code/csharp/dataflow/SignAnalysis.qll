/**
 * Provides sign analysis to determine whether expression are always positive
 * or negative.
 *
 * The analysis is implemented as an abstract interpretation over the
 * three-valued domain `{negative, zero, positive}`.
 */

private import csharp
private import Ssa
private import SsaReadPosition
private import ConstantUtils
private import RangeUtils
private import semmle.code.csharp.controlflow.Guards as G
private import Linq.Helpers as Linq

private class Guard = G::Guard;

private newtype TSign =
  TNeg() or
  TZero() or
  TPos()

private class Sign extends TSign {
  string toString() {
    result = "-" and this = TNeg()
    or
    result = "0" and this = TZero()
    or
    result = "+" and this = TPos()
  }

  Sign inc() {
    this = TNeg() and result = TNeg()
    or
    this = TNeg() and result = TZero()
    or
    this = TZero() and result = TPos()
    or
    this = TPos() and result = TPos()
  }

  Sign dec() { result.inc() = this }

  Sign neg() {
    this = TNeg() and result = TPos()
    or
    this = TZero() and result = TZero()
    or
    this = TPos() and result = TNeg()
  }

  Sign bitnot() {
    this = TNeg() and result = TPos()
    or
    this = TNeg() and result = TZero()
    or
    this = TZero() and result = TNeg()
    or
    this = TPos() and result = TNeg()
  }

  Sign add(Sign s) {
    this = TZero() and result = s
    or
    s = TZero() and result = this
    or
    this = s and this = result
    or
    this = TPos() and s = TNeg()
    or
    this = TNeg() and s = TPos()
  }

  Sign mul(Sign s) {
    result = TZero() and this = TZero()
    or
    result = TZero() and s = TZero()
    or
    result = TNeg() and this = TPos() and s = TNeg()
    or
    result = TNeg() and this = TNeg() and s = TPos()
    or
    result = TPos() and this = TPos() and s = TPos()
    or
    result = TPos() and this = TNeg() and s = TNeg()
  }

  Sign div(Sign s) {
    result = TZero() and this = TZero() and s = TNeg()
    or
    result = TZero() and this = TZero() and s = TPos()
    or
    result = TNeg() and this = TPos() and s = TNeg()
    or
    result = TNeg() and this = TNeg() and s = TPos()
    or
    result = TPos() and this = TPos() and s = TPos()
    or
    result = TPos() and this = TNeg() and s = TNeg()
  }

  Sign rem(Sign s) {
    result = TZero() and s = TNeg()
    or
    result = TZero() and s = TPos()
    or
    result = this and s = TNeg()
    or
    result = this and s = TPos()
  }

  Sign bitand(Sign s) {
    result = TZero() and this = TZero()
    or
    result = TZero() and s = TZero()
    or
    result = TZero() and this = TPos()
    or
    result = TZero() and s = TPos()
    or
    result = TNeg() and this = TNeg() and s = TNeg()
    or
    result = TPos() and this = TNeg() and s = TPos()
    or
    result = TPos() and this = TPos() and s = TNeg()
    or
    result = TPos() and this = TPos() and s = TPos()
  }

  Sign bitor(Sign s) {
    result = TZero() and this = TZero() and s = TZero()
    or
    result = TNeg() and this = TNeg()
    or
    result = TNeg() and s = TNeg()
    or
    result = TPos() and this = TPos() and s = TZero()
    or
    result = TPos() and this = TZero() and s = TPos()
    or
    result = TPos() and this = TPos() and s = TPos()
  }

  Sign bitxor(Sign s) {
    result = TZero() and this = s
    or
    result = this and s = TZero()
    or
    result = s and this = TZero()
    or
    result = TPos() and this = TPos() and s = TPos()
    or
    result = TNeg() and this = TNeg() and s = TPos()
    or
    result = TNeg() and this = TPos() and s = TNeg()
    or
    result = TPos() and this = TNeg() and s = TNeg()
  }

  Sign lshift(Sign s) {
    result = TZero() and this = TZero()
    or
    result = this and s = TZero()
    or
    this != TZero() and s != TZero()
  }

  Sign rshift(Sign s) {
    result = TZero() and this = TZero()
    or
    result = this and s = TZero()
    or
    result = TNeg() and this = TNeg()
    or
    result != TNeg() and this = TPos() and s != TZero()
  }
}

/** Gets the sign of `e` if this can be directly determined. */
private Sign certainExprSign(Expr e) {
  exists(int i | e.(ConstantIntegerExpr).getIntValue() = i |
    i < 0 and result = TNeg()
    or
    i = 0 and result = TZero()
    or
    i > 0 and result = TPos()
  )
  or
  not exists(e.(ConstantIntegerExpr).getIntValue()) and
  (
    exists(float f |
      f = e.(LongLiteral).getValue().toFloat() or
      f = e.(RealLiteral).getValue().toFloat()
    |
      f < 0 and result = TNeg()
      or
      f = 0 and result = TZero()
      or
      f > 0 and result = TPos()
    )
    or
    exists(string charlit | charlit = e.(CharLiteral).getValue() |
      if charlit = "\\0" or charlit = "\\u0000" then result = TZero() else result = TPos()
    )
    or
    containerSizeAccess(e.(PropertyAccess)) and
    (result = TPos() or result = TZero())
    or
    e instanceof Linq::CountCall and
    (result = TPos() or result = TZero())
  )
}

private predicate containerSizeAccess(PropertyAccess pa) {
  propertyOverrides(pa.getTarget(), "System.Collections.Generic.ICollection<>", "Count") or
  propertyOverrides(pa.getTarget(), "System.Collections.Generic.IReadOnlyCollection<>", "Count") or
  propertyOverrides(pa.getTarget(), "System.Collections.ICollection", "Count") or
  propertyOverrides(pa.getTarget(), "System.String", "Length") or
  propertyOverrides(pa.getTarget(), "System.Array", "Length")
}

private class NumericOrCharType extends Type {
  NumericOrCharType() {
    this instanceof CharType or
    this instanceof IntegralType or
    this instanceof FloatingPointType or
    this instanceof DecimalType
  }
}

/** Holds if the sign of `e` is too complicated to determine. */
private predicate unknownSign(Expr e) {
  not exists(e.(ConstantIntegerExpr).getIntValue()) and
  (
    exists(IntegerLiteral lit | lit = e and not exists(lit.getValue().toInt()))
    or
    exists(LongLiteral lit | lit = e and not exists(lit.getValue().toFloat()))
    or
    exists(CastExpr cast, Type fromtyp |
      cast = e and
      fromtyp = cast.getExpr().getType() and
      not fromtyp instanceof NumericOrCharType
    )
    or
    e instanceof ArrayAccess and e.getType() instanceof NumericOrCharType
    or
    e instanceof MethodAccess and e.getType() instanceof NumericOrCharType
    or
    e instanceof ObjectCreation and e.getType() instanceof NumericOrCharType
  )
}

/**
 * Holds if `lowerbound` is a lower bound for `v` at `pos`. This is restricted
 * to only include bounds for which we might determine a sign.
 */
private predicate lowerBound(Expr lowerbound, Definition v, SsaReadPosition pos, boolean isStrict) {
  exists(boolean testIsTrue, RelationalOperation comp |
    pos.hasReadOfVar(v) and
    guardControlsSsaRead(comp, pos, testIsTrue) and
    not unknownSign(lowerbound)
  |
    testIsTrue = true and
    comp.getLesserOperand() = lowerbound and
    comp.getGreaterOperand() = ssaRead(v, 0) and
    (if comp.isStrict() then isStrict = true else isStrict = false)
    or
    testIsTrue = false and
    comp.getGreaterOperand() = lowerbound and
    comp.getLesserOperand() = ssaRead(v, 0) and
    (if comp.isStrict() then isStrict = false else isStrict = true)
  )
}

/**
 * Holds if `upperbound` is an upper bound for `v` at `pos`. This is restricted
 * to only include bounds for which we might determine a sign.
 */
private predicate upperBound(Expr upperbound, Definition v, SsaReadPosition pos, boolean isStrict) {
  exists(boolean testIsTrue, RelationalOperation comp |
    pos.hasReadOfVar(v) and
    guardControlsSsaRead(comp, pos, testIsTrue) and
    not unknownSign(upperbound)
  |
    testIsTrue = true and
    comp.getGreaterOperand() = upperbound and
    comp.getLesserOperand() = ssaRead(v, 0) and
    (if comp.isStrict() then isStrict = true else isStrict = false)
    or
    testIsTrue = false and
    comp.getLesserOperand() = upperbound and
    comp.getGreaterOperand() = ssaRead(v, 0) and
    (if comp.isStrict() then isStrict = false else isStrict = true)
  )
}

/**
 * Holds if `eqbound` is an equality/inequality for `v` at `pos`. This is
 * restricted to only include bounds for which we might determine a sign. The
 * boolean `isEq` gives the polarity:
 *  - `isEq = true` : `v = eqbound`
 *  - `isEq = false` : `v != eqbound`
 */
private predicate eqBound(Expr eqbound, Definition v, SsaReadPosition pos, boolean isEq) {
  exists(Guard guard, boolean testIsTrue, boolean polarity |
    pos.hasReadOfVar(v) and
    guardControlsSsaRead(guard, pos, testIsTrue) and
    guard.isEquality(eqbound, ssaRead(v, 0), polarity) and
    isEq = polarity.booleanXor(testIsTrue).booleanNot() and
    not unknownSign(eqbound)
  )
}

/**
 * Holds if `bound` is a bound for `v` at `pos` that needs to be positive in
 * order for `v` to be positive.
 */
private predicate posBound(Expr bound, Definition v, SsaReadPosition pos, boolean isStrict) {
  lowerBound(bound, v, pos, isStrict)
  or
  eqBound(bound, v, pos, true) and isStrict = false
}

/**
 * Holds if `bound` is a bound for `v` at `pos` that needs to be negative in
 * order for `v` to be negative.
 */
private predicate negBound(Expr bound, Definition v, SsaReadPosition pos, boolean isStrict) {
  upperBound(bound, v, pos, isStrict)
  or
  eqBound(bound, v, pos, true) and isStrict = false
}

/**
 * Holds if `bound` is a bound for `v` at `pos` that can restrict whether `v`
 * can be zero.
 */
private predicate zeroBound(Expr bound, Definition v, SsaReadPosition pos) {
  lowerBound(bound, v, pos, _) or
  upperBound(bound, v, pos, _) or
  eqBound(bound, v, pos, _)
}

/** Holds if `bound` allows `v` to be positive at `pos`. */
private predicate posBoundOk(Expr bound, Definition v, SsaReadPosition pos) {
  posBound(bound, v, pos, _) and TPos() = exprSign(bound)
  or
  posBound(bound, v, pos, true) and TZero() = exprSign(bound)
}

/** Holds if `bound` allows `v` to be negative at `pos`. */
private predicate negBoundOk(Expr bound, Definition v, SsaReadPosition pos) {
  negBound(bound, v, pos, _) and TNeg() = exprSign(bound)
  or
  negBound(bound, v, pos, true) and TZero() = exprSign(bound)
}

/** Holds if `bound` allows `v` to be zero at `pos`. */
private predicate zeroBoundOk(Expr bound, Definition v, SsaReadPosition pos) {
  lowerBound(bound, v, pos, _) and TNeg() = exprSign(bound)
  or
  lowerBound(bound, v, pos, false) and TZero() = exprSign(bound)
  or
  upperBound(bound, v, pos, _) and TPos() = exprSign(bound)
  or
  upperBound(bound, v, pos, false) and TZero() = exprSign(bound)
  or
  eqBound(bound, v, pos, true) and TZero() = exprSign(bound)
  or
  eqBound(bound, v, pos, false) and TZero() != exprSign(bound)
}

/**
 * Holds if there is a bound that might restrict whether `v` has the sign `s`
 * at `pos`.
 */
predicate hasGuard(Definition v, SsaReadPosition pos, Sign s) {
  s = TPos() and posBound(_, v, pos, _)
  or
  s = TNeg() and negBound(_, v, pos, _)
  or
  s = TZero() and zeroBound(_, v, pos)
}

// pragma[noinline]
// private Sign guardedSsaSign(Definition v, SsaReadPosition pos) {
//   //result = ssaDefSign(v) and
//   pos.hasReadOfVar(v) and
//   hasGuard(v, pos, result)
// }
pragma[noinline]
private Sign unguardedSsaSign(Definition v, SsaReadPosition pos) {
  result = ssaDefSign(v) and
  pos.hasReadOfVar(v) and
  not hasGuard(v, pos, result)
}

private Sign guardedSsaSignOk(Definition v, SsaReadPosition pos) {
  result = TPos() and
  forex(Expr bound | posBound(bound, v, pos, _) | posBoundOk(bound, v, pos))
  or
  result = TNeg() and
  forex(Expr bound | negBound(bound, v, pos, _) | negBoundOk(bound, v, pos))
  or
  result = TZero() and
  forex(Expr bound | zeroBound(bound, v, pos) | zeroBoundOk(bound, v, pos))
}

/** Gets a possible sign for `v` at `pos`. */
Sign ssaSign(Definition v, SsaReadPosition pos) {
  result = unguardedSsaSign(v, pos)
  or
  //result = guardedSsaSign(v, pos) and
  result = guardedSsaSignOk(v, pos)
}

/** Gets a possible sign for `v`. */
pragma[nomagic]
private Sign ssaDefSign(Definition v) {
  exists(Element def | def = v.(ExplicitDefinition).getADefinition().getElement() |
    result = exprSign(def.(Assignment).getRValue())
    or
    result = exprSign(def.(PostIncrExpr).getOperand()).inc()
    or
    result = exprSign(def.(PreIncrExpr).getOperand()).inc()
    or
    result = exprSign(def.(PostDecrExpr).getOperand()).dec()
    or
    result = exprSign(def.(PreDecrExpr).getOperand()).dec()
    or
    exists(AssignOperation a | a = def and result = exprSign(a))
  )
  or
  result = fieldSign(v.(ImplicitDefinition).getSourceVariable().getAssignable())
  or
  //   exists(Parameter p | v.(ImplicitEntryDefinition).getSourceVariable().getAssignable() = p)
  //   or
  exists(PhiNode phi, Definition inp, SsaReadPositionPhiInputEdge edge |
    v = phi and
    edge.phiInput(phi, inp) and
    result = ssaSign(inp, edge)
  )
}

/** Gets a possible sign for `f`. */
private Sign fieldSign(Field f) {
  result = exprSign(f.getAnAssignedValue())
  or
  exists(PostIncrExpr inc | inc.getOperand() = f.getAnAccess() and result = fieldSign(f).inc())
  or
  exists(PreIncrExpr inc | inc.getOperand() = f.getAnAccess() and result = fieldSign(f).inc())
  or
  exists(PostDecrExpr inc | inc.getOperand() = f.getAnAccess() and result = fieldSign(f).dec())
  or
  exists(PreDecrExpr inc | inc.getOperand() = f.getAnAccess() and result = fieldSign(f).dec())
  or
  exists(AssignOperation a | a.getLValue() = f.getAnAccess() | result = exprSign(a))
  or
  f.fromSource() and not exists(f.getInitializer()) and result = TZero()
}

/** Gets a possible sign for `e`. */
cached
private Sign exprSign(Expr e) {
  result = certainExprSign(e)
  or
  not exists(certainExprSign(e)) and
  (
    unknownSign(e)
    or
    exists(Definition v | v.getARead() = e |
      result = ssaSign(v, any(SsaReadPositionBlock bb | bb.getBlock().getANode().getElement() = e))
      or
      not exists(SsaReadPositionBlock bb | bb.getBlock().getANode().getElement() = e) and
      result = ssaDefSign(v)
    )
    or
    exists(FieldAccess fa | fa = e |
      not exists(Definition v | v.getARead() = fa) and
      result = fieldSign(fa.getTarget())
    )
    or
    exists(VariableAccess va | va = e |
      not exists(Definition v | v.getARead() = va) and
      not va instanceof FieldAccess
    )
    or
    result = exprSign(e.(AssignExpr).getRValue())
    or
    result = exprSign(e.(UnaryPlusExpr).getOperand())
    or
    result = exprSign(e.(PostIncrExpr).getOperand())
    or
    result = exprSign(e.(PostDecrExpr).getOperand())
    or
    result = exprSign(e.(PreIncrExpr).getOperand()).inc()
    or
    result = exprSign(e.(PreDecrExpr).getOperand()).dec()
    or
    result = exprSign(e.(UnaryMinusExpr).getOperand()).neg()
    or
    result = exprSign(e.(ComplementExpr).getOperand()).bitnot()
    or
    exists(DivExpr div |
      div = e and
      result = exprSign(div.getLeftOperand()) and
      result != TZero()
    |
      div.getRightOperand().(RealLiteral).getValue().toFloat() = 0
    )
    or
    exists(Sign s1, Sign s2 | binaryOpSigns(e, s1, s2) |
      (e instanceof AssignAddExpr or e instanceof AddExpr) and
      result = s1.add(s2)
      or
      (e instanceof AssignSubExpr or e instanceof SubExpr) and
      result = s1.add(s2.neg())
      or
      (e instanceof AssignMulExpr or e instanceof MulExpr) and
      result = s1.mul(s2)
      or
      (e instanceof AssignDivExpr or e instanceof DivExpr) and
      result = s1.div(s2)
      or
      (e instanceof AssignRemExpr or e instanceof RemExpr) and
      result = s1.rem(s2)
      or
      (e instanceof AssignAndExpr or e instanceof BitwiseAndExpr) and
      result = s1.bitand(s2)
      or
      (e instanceof AssignOrExpr or e instanceof BitwiseOrExpr) and
      result = s1.bitor(s2)
      or
      (e instanceof AssignXorExpr or e instanceof BitwiseXorExpr) and
      result = s1.bitxor(s2)
      or
      (e instanceof AssignLShiftExpr or e instanceof LShiftExpr) and
      result = s1.lshift(s2)
      or
      (e instanceof AssignRShiftExpr or e instanceof RShiftExpr) and
      result = s1.rshift(s2)
    )
    or
    result = exprSign(e.(ConditionalExpr).getThen())
    or
    result = exprSign(e.(ConditionalExpr).getElse())
    or
    result = exprSign(e.(SwitchExpr).getACase().getBody())
    or
    result = exprSign(e.(CastExpr).getExpr())
  )
}

private Sign binaryOpLhsSign(Expr e) {
  result = exprSign(e.(BinaryOperation).getLeftOperand()) or
  result = exprSign(e.(AssignOperation).getLValue())
}

private Sign binaryOpRhsSign(Expr e) {
  result = exprSign(e.(BinaryOperation).getRightOperand()) or
  result = exprSign(e.(AssignOperation).getRValue())
}

pragma[noinline]
private predicate binaryOpSigns(Expr e, Sign lhs, Sign rhs) {
  lhs = binaryOpLhsSign(e) and
  rhs = binaryOpRhsSign(e)
}

/** Holds if `e` can be positive and cannot be negative. */
predicate positive(Expr e) {
  exprSign(e) = TPos() and
  not exprSign(e) = TNeg()
}

/** Holds if `e` can be negative and cannot be positive. */
predicate negative(Expr e) {
  exprSign(e) = TNeg() and
  not exprSign(e) = TPos()
}

/** Holds if `e` is strictly positive. */
predicate strictlyPositive(Expr e) {
  exprSign(e) = TPos() and
  not exprSign(e) = TNeg() and
  not exprSign(e) = TZero()
}

/** Holds if `e` is strictly negative. */
predicate strictlyNegative(Expr e) {
  exprSign(e) = TNeg() and
  not exprSign(e) = TPos() and
  not exprSign(e) = TZero()
}
