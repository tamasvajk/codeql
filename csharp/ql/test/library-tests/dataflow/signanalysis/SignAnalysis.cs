using System;
using System.Linq;

class SignAnalysis
{
    void Test1(bool b)
    {
        {
            int i = 0;
            System.Console.WriteLine(i);
        }
        {
            int i = 1 + 1;
            System.Console.WriteLine(i);
        }
        {
            int i = -1;
            System.Console.WriteLine(i);
        }
        {
            int i = 0;
            i++;
            System.Console.WriteLine(i);
        }
        {
            int i = 0;
            i--;
            System.Console.WriteLine(i);
        }
        {
            int i = 0;
            ++i;
            System.Console.WriteLine(i);
        }
        {
            int i = 0;
            --i;
            System.Console.WriteLine(i);
        }
        {
            int i = 0;
            --i;
            var j = i;
            System.Console.WriteLine(j);
        }
        {
            int i = 1;
            if (b)
            {
                i++;
            }
            System.Console.WriteLine(i);
        }
        {
            int i = 1;
            if (b)
            {
                i++;
            }
            else
            {
                i--;
            }
            System.Console.WriteLine(i);
        }
    }

    void Test2()
    {
        var i = 1;
        i = i * i;
        System.Console.WriteLine(i);
        i = i * -i;
        System.Console.WriteLine(i);
        i = -i * -i;
        System.Console.WriteLine(i);
        i = -i / -i;
        System.Console.WriteLine(i);
        i = -i % i;
        System.Console.WriteLine(i);
        i = -i;
        System.Console.WriteLine(i);
        i = +i;
        System.Console.WriteLine(i);
    }
    void Test3()
    {
        var f = 4.2f;
        System.Console.WriteLine(f);
        var d = 4.2;
        System.Console.WriteLine(d);
        var de = 4.2m;
        System.Console.WriteLine(de);
        var c = 'a';
        System.Console.WriteLine(c);
    }

    int f1;
    int f2 = 1;

    void TestField()
    {
        f1++;
        System.Console.WriteLine(f1);
        System.Console.WriteLine(f2);
        var j = f1;
        System.Console.WriteLine(j);
    }

    void Test4()
    {
        var i = new Int32();
        i++;
        System.Console.WriteLine(i);

        var ints = new int[] { 1, 2, 3 };
        System.Console.WriteLine(ints[0]);
    }

    int SomeValue() { return (new System.Random()).Next(0, 1000); }
    void Test5()
    {
        var i = SomeValue();
        if (i < 0)
        {
            System.Console.WriteLine(i);
            return;
        }

        System.Console.WriteLine(i); // positive, but not recognized

        if (i != 0)
        {
            System.Console.WriteLine(i); // strictly positive, but not recognized
        }
    }

    void Test5a()
    {
        var i = SomeValue();
        if (i == 1)
        {
            System.Console.WriteLine(i);
            return;
        }
    }

    void Test5b()
    {
        var i = SomeValue();
        if (i > 0)
        {
            System.Console.WriteLine(i);
            return;
        }
    }

    void Test6(int i)
    {
        if (i < 0)
        {
            System.Console.WriteLine(i);
        }
    }

    void Test7(int[] p)
    {
        var ints = p;
        System.Console.WriteLine(ints.Length); // positive
        ints = new int[] { 1, 2, 3 };
        System.Console.WriteLine(ints.Length); // 3, so strictly positive
        System.Console.WriteLine(ints.Count());
        System.Console.WriteLine(ints.Count(i => i > 1));

        var s = "abc";
        System.Console.WriteLine(s.Length);

        System.Console.WriteLine(int.MaxValue);
        System.Console.WriteLine(int.MinValue);
    }
}

// semmle-extractor-options: /r:System.Linq.dll