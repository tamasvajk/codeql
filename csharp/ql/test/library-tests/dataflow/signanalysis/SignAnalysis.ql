import csharp
import semmle.code.csharp.dataflow.SignAnalysis
import semmle.code.csharp.dataflow.SsaReadPosition
import semmle.code.csharp.dataflow.SSA
import semmle.code.csharp.dataflow.RangeUtils

private predicate variableAcc(
  VariableAccess va, Ssa::Definition v, Expr guard, Expr bound, string boundSign, string gSign
) {
  va.fromSource() and
  va.getEnclosingCallable().getDeclaringType().getName() = "SignAnalysis" and
  va.getLocation().getStartLine() = 85 and
  v.getARead() = va and
  exists(SsaReadPosition pos, Sign s, Sign gs |
    pos.hasReadOfVar(v) and
    guardControlsSsaRead(guard, pos, true) and
    upperBound(bound, v, pos, false) and
    s = exprSign(bound) and
    boundSign = s.toString() and
    negBoundOk(bound, v, pos) and
    gs = ssaSign(v, pos) and
    gSign = gs.toString() and
    negative(va)
  )
}

string getASignString(Expr e) {
  positive(e) and
  result = "positive"
  or
  negative(e) and
  result = "negative"
  or
  strictlyPositive(e) and
  result = "strictlyPositive"
  or
  strictlyNegative(e) and
  result = "strictlyNegative"
}

from Expr e
select e, strictconcat(string s | s = getASignString(e) | s, " ")
