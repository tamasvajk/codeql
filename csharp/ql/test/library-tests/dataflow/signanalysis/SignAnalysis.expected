| SignAnalysis.cs:13:21:13:21 | 1 | positive strictlyPositive |
| SignAnalysis.cs:13:21:13:25 | ... + ... | positive strictlyPositive |
| SignAnalysis.cs:13:25:13:25 | 1 | positive strictlyPositive |
| SignAnalysis.cs:14:38:14:38 | access to local variable i | positive strictlyPositive |
| SignAnalysis.cs:17:21:17:22 | -... | negative strictlyNegative |
| SignAnalysis.cs:17:22:17:22 | 1 | positive strictlyPositive |
| SignAnalysis.cs:18:38:18:38 | access to local variable i | negative strictlyNegative |
| SignAnalysis.cs:23:38:23:38 | access to local variable i | positive strictlyPositive |
| SignAnalysis.cs:28:38:28:38 | access to local variable i | negative strictlyNegative |
| SignAnalysis.cs:32:13:32:15 | ++... | positive strictlyPositive |
| SignAnalysis.cs:33:38:33:38 | access to local variable i | positive strictlyPositive |
| SignAnalysis.cs:37:13:37:15 | --... | negative strictlyNegative |
| SignAnalysis.cs:38:38:38:38 | access to local variable i | negative strictlyNegative |
| SignAnalysis.cs:42:13:42:15 | --... | negative strictlyNegative |
| SignAnalysis.cs:43:21:43:21 | access to local variable i | negative strictlyNegative |
| SignAnalysis.cs:44:38:44:38 | access to local variable j | negative strictlyNegative |
| SignAnalysis.cs:47:21:47:21 | 1 | positive strictlyPositive |
| SignAnalysis.cs:50:17:50:17 | access to local variable i | positive strictlyPositive |
| SignAnalysis.cs:50:17:50:19 | ...++ | positive strictlyPositive |
| SignAnalysis.cs:52:38:52:38 | access to local variable i | positive strictlyPositive |
| SignAnalysis.cs:55:21:55:21 | 1 | positive strictlyPositive |
| SignAnalysis.cs:58:17:58:17 | access to local variable i | positive strictlyPositive |
| SignAnalysis.cs:58:17:58:19 | ...++ | positive strictlyPositive |
| SignAnalysis.cs:62:17:62:17 | access to local variable i | positive strictlyPositive |
| SignAnalysis.cs:62:17:62:19 | ...-- | positive strictlyPositive |
| SignAnalysis.cs:64:38:64:38 | access to local variable i | positive |
| SignAnalysis.cs:70:17:70:17 | 1 | positive strictlyPositive |
| SignAnalysis.cs:71:9:71:17 | ... = ... | positive strictlyPositive |
| SignAnalysis.cs:71:13:71:13 | access to local variable i | positive strictlyPositive |
| SignAnalysis.cs:71:13:71:17 | ... * ... | positive strictlyPositive |
| SignAnalysis.cs:71:17:71:17 | access to local variable i | positive strictlyPositive |
| SignAnalysis.cs:72:34:72:34 | access to local variable i | positive strictlyPositive |
| SignAnalysis.cs:73:9:73:18 | ... = ... | negative strictlyNegative |
| SignAnalysis.cs:73:13:73:13 | access to local variable i | positive strictlyPositive |
| SignAnalysis.cs:73:13:73:18 | ... * ... | negative strictlyNegative |
| SignAnalysis.cs:73:17:73:18 | -... | negative strictlyNegative |
| SignAnalysis.cs:73:18:73:18 | access to local variable i | positive strictlyPositive |
| SignAnalysis.cs:74:34:74:34 | access to local variable i | negative strictlyNegative |
| SignAnalysis.cs:75:9:75:19 | ... = ... | positive strictlyPositive |
| SignAnalysis.cs:75:13:75:14 | -... | positive strictlyPositive |
| SignAnalysis.cs:75:13:75:19 | ... * ... | positive strictlyPositive |
| SignAnalysis.cs:75:14:75:14 | access to local variable i | negative strictlyNegative |
| SignAnalysis.cs:75:18:75:19 | -... | positive strictlyPositive |
| SignAnalysis.cs:75:19:75:19 | access to local variable i | negative strictlyNegative |
| SignAnalysis.cs:76:34:76:34 | access to local variable i | positive strictlyPositive |
| SignAnalysis.cs:77:9:77:19 | ... = ... | positive strictlyPositive |
| SignAnalysis.cs:77:13:77:14 | -... | negative strictlyNegative |
| SignAnalysis.cs:77:13:77:19 | ... / ... | positive strictlyPositive |
| SignAnalysis.cs:77:14:77:14 | access to local variable i | positive strictlyPositive |
| SignAnalysis.cs:77:18:77:19 | -... | negative strictlyNegative |
| SignAnalysis.cs:77:19:77:19 | access to local variable i | positive strictlyPositive |
| SignAnalysis.cs:78:34:78:34 | access to local variable i | positive strictlyPositive |
| SignAnalysis.cs:79:9:79:18 | ... = ... | negative |
| SignAnalysis.cs:79:13:79:14 | -... | negative strictlyNegative |
| SignAnalysis.cs:79:13:79:18 | ... % ... | negative |
| SignAnalysis.cs:79:14:79:14 | access to local variable i | positive strictlyPositive |
| SignAnalysis.cs:79:18:79:18 | access to local variable i | positive strictlyPositive |
| SignAnalysis.cs:80:34:80:34 | access to local variable i | negative |
| SignAnalysis.cs:81:9:81:14 | ... = ... | positive |
| SignAnalysis.cs:81:13:81:14 | -... | positive |
| SignAnalysis.cs:81:14:81:14 | access to local variable i | negative |
| SignAnalysis.cs:82:34:82:34 | access to local variable i | positive |
| SignAnalysis.cs:83:9:83:14 | ... = ... | positive |
| SignAnalysis.cs:83:13:83:14 | +... | positive |
| SignAnalysis.cs:83:14:83:14 | access to local variable i | positive |
| SignAnalysis.cs:84:34:84:34 | access to local variable i | positive |
| SignAnalysis.cs:88:17:88:20 | 4.2 | positive strictlyPositive |
| SignAnalysis.cs:89:34:89:34 | access to local variable f | positive strictlyPositive |
| SignAnalysis.cs:90:17:90:19 | 4.2 | positive strictlyPositive |
| SignAnalysis.cs:91:34:91:34 | access to local variable d | positive strictlyPositive |
| SignAnalysis.cs:92:18:92:21 | 4.2 | positive strictlyPositive |
| SignAnalysis.cs:93:34:93:35 | access to local variable de | positive strictlyPositive |
| SignAnalysis.cs:94:17:94:19 | a | positive strictlyPositive |
| SignAnalysis.cs:95:34:95:34 | access to local variable c | positive strictlyPositive |
| SignAnalysis.cs:99:9:99:10 | access to field f2 | positive strictlyPositive |
| SignAnalysis.cs:99:9:99:14 | ... = ... | positive strictlyPositive |
| SignAnalysis.cs:99:14:99:14 | 1 | positive strictlyPositive |
| SignAnalysis.cs:103:9:103:10 | access to field f1 | positive |
| SignAnalysis.cs:103:9:103:12 | ...++ | positive |
| SignAnalysis.cs:104:34:104:35 | access to field f1 | positive strictlyPositive |
| SignAnalysis.cs:105:34:105:35 | access to field f2 | positive strictlyPositive |
| SignAnalysis.cs:106:17:106:18 | access to field f1 | positive strictlyPositive |
| SignAnalysis.cs:107:34:107:34 | access to local variable j | positive strictlyPositive |
| SignAnalysis.cs:114:34:114:34 | access to local variable i | positive strictlyPositive |
| SignAnalysis.cs:116:20:116:40 | 3 | positive strictlyPositive |
| SignAnalysis.cs:116:32:116:32 | 1 | positive strictlyPositive |
| SignAnalysis.cs:116:35:116:35 | 2 | positive strictlyPositive |
| SignAnalysis.cs:116:38:116:38 | 3 | positive strictlyPositive |
| SignAnalysis.cs:120:60:120:63 | 1000 | positive strictlyPositive |
| SignAnalysis.cs:141:18:141:18 | 1 | positive strictlyPositive |
| SignAnalysis.cs:169:34:169:44 | access to property Length | positive |
| SignAnalysis.cs:170:16:170:36 | 3 | positive strictlyPositive |
| SignAnalysis.cs:170:28:170:28 | 1 | positive strictlyPositive |
| SignAnalysis.cs:170:31:170:31 | 2 | positive strictlyPositive |
| SignAnalysis.cs:170:34:170:34 | 3 | positive strictlyPositive |
| SignAnalysis.cs:171:34:171:44 | access to property Length | positive strictlyPositive |
| SignAnalysis.cs:172:34:172:45 | call to method Count | positive |
| SignAnalysis.cs:173:34:173:55 | call to method Count | positive |
| SignAnalysis.cs:173:54:173:54 | 1 | positive strictlyPositive |
| SignAnalysis.cs:176:34:176:41 | access to property Length | positive |
| SignAnalysis.cs:178:34:178:45 | access to constant MaxValue | positive strictlyPositive |
| SignAnalysis.cs:179:34:179:45 | access to constant MinValue | negative strictlyNegative |
| SignAnalysis.cs:185:18:185:19 | -... | negative strictlyNegative |
| SignAnalysis.cs:185:19:185:19 | 1 | positive strictlyPositive |
| SignAnalysis.cs:202:18:202:18 | 1 | positive strictlyPositive |
| SignAnalysis.cs:211:18:211:19 | -... | negative strictlyNegative |
| SignAnalysis.cs:211:19:211:19 | 1 | positive strictlyPositive |
